hyacinth/session.py

```diff
@@ -236,26 +241,103 @@ def delete_folder(self, id, **kwargs):
         return self.__delete_resource(url, **kwargs)
 
-    def upload_document(self, name, parent_id, parent_type, document):
+    def upload_document(self, name, parent_id, parent_type, document, progress_update=lambda *args: None):
         """POST a new Document, PUT the data, and PATCH Document as fully_uploaded."""
+        file_size = os.path.getsize(document)
+        if file_size > PART_SIZE:
+            return self.upload_multipart_document(
+                name,
+                parent_id,
+                parent_type,
+                document,
+                progress_update,
+            )
+
+        with open(document, "rb") as f:
+            post_url = Session.__make_url("documents")
+            clio_document = self.__post_resource(
+                post_url,
+                params={"fields": "id,latest_document_version{uuid,put_url,put_headers}"},
+                json={
+                    "data": {"name": name, "parent": {"id": parent_id, "type": parent_type}}
+                },
+            )
+
+            put_url = clio_document["data"]["latest_document_version"]["put_url"]
+            put_headers = clio_document["data"]["latest_document_version"]["put_headers"]
+
+            headers_map = {}
+            for header in put_headers:
+                headers_map[header["name"]] = header["value"]
+
+            # We actually DON'T want to use the authenticated client here
+            requests.put(put_url, headers=headers_map, data=f, timeout=5)
+
+            patch_url = self.__make_url(f"documents/{clio_document['data']['id']}")
+            patch_resp = self.__patch_resource(
+                patch_url,
+                params={"fields": "id,name,latest_document_version{fully_uploaded}"},
+                json={
+                    "data": {
+                        "uuid": clio_document["data"]["latest_document_version"]["uuid"],
+                        "fully_uploaded": True,
+                    }
+                },
+            )
+
+            return patch_resp
+
+    def upload_multipart_document(
+            self, name, parent_id, parent_type, document, progress_update
+    ):
+        """Upload a new Document to Clio via the multipart upload feature."""
+        with open(document, "rb") as f:
+            file_size = os.path.getsize(document)
+            parts = []
+            for i in range(0, file_size, PART_SIZE):
+                start_offset = i
+                end_offset = min(file_size, i + PART_SIZE)
+                part = f.read(end_offset - start_offset)
+                parts.append((start_offset, end_offset, part))
+                progress_update()
+            content_md5 = hashlib.md5(part).digest()  # nosec
+            content_md5_str = base64.b64encode(content_md5).decode('utf-8')
+
+        multiparts = []
+        for idx, part in enumerate(parts, start=1):
+            multiparts.append({
+                "part_number": idx,
+                "content_length": len(part[2]),
```

Comment:
What's in `part[2]`?

Reply:
`part[2]` here is the binary data "part". This is not very well documented, it's true-- good shout.